### PR TITLE
Fix digamma(0) to return -inf matching scipy

### LINKED ISSUE #37763: `jax.scipy.special.digamma(0.0)` returns `nan` — should be `-inf`
### Description

`jax.scipy.special.digamma` returns `nan` at `x = 0`. The correct value is `-inf` because digamma has a simple pole at the origin. SciPy handles this correctly.

This is a boundary-value correctness bug, not a precision issue. The digamma function is well-defined near zero (it diverges to −∞), and any implementation should return `-inf` rather than `nan` for `x = 0`.

I ran into this while computing gradients of a Dirichlet log-likelihood. The gradient involves `digamma(alpha_k)` and when one concentration parameter is very small, the result should be a large negative number — not `nan` which poisons the entire gradient vector.

**Secondary issue**: because XLA flushes denormal inputs to zero (DAZ), subnormal float32 inputs also trigger the `x = 0` code path, so `digamma(subnormal)` returns `nan` too. The correct value for small positive `x` is approximately `−1/x`, which would be a large negative number or `-inf` for the smallest subnormals.

## Reproducer

```python
from jax.scipy.special import digamma as jax_digamma
from scipy.special import digamma as sp_digamma
import jax.numpy as jnp
import numpy as np

test_vals = [0.0, np.nextafter(np.finfo(np.float32).tiny, 0.0), 1e-40, 0.5, 1.0]

for v in test_vals:
    sp  = sp_digamma(v)
    jax = float(jax_digamma(jnp.float32(v)))
    print(f"digamma({v:.2e}):  scipy={sp:.4g}   jax={jax}")
```

## Expected output

```
digamma(0.00e+00):  scipy=-inf        jax=-inf
digamma(1.18e-38):  scipy=-8.507e+37  jax=-8.507e+37
digamma(1.00e-40):  scipy=-1e+40      jax=-inf
digamma(5.00e-01):  scipy=-1.964      jax=-1.964
digamma(1.00e+00):  scipy=-0.5772     jax=-0.5772
```

## Actual output

```
digamma(0.00e+00):  scipy=-inf        jax=nan    ← wrong
digamma(1.18e-38):  scipy=-8.507e+37  jax=nan    ← wrong
digamma(1.00e-40):  scipy=-1e+40      jax=nan    ← wrong
digamma(5.00e-01):  scipy=-1.964      jax=-1.964 ← ok
digamma(1.00e+00):  scipy=-0.5772     jax=-0.5772 ← ok
```

## Notes

- The `x = 0` case is the primary bug. The subnormal failures are a consequence of it combined with XLA's DAZ mode.
- `jnp.log(0.0)` correctly returns `-inf`, so the issue is specific to digamma's boundary handling, not a general `-inf` problem in JAX.
- Normal-range inputs work correctly.

### System info (python version, jaxlib version, accelerator, etc.)

jax:     0.10.1.dev20260513+64f6e229b
jaxlib:  0.10.0
numpy:   2.4.4
scipy:   1.17.1
python:  3.12.3
OS:      Linux-6.14.0-37-generic-x86_64-with-glibc2.39
backend: cpu

### PR DESCRIPTION
**Pull request by adityasingh2400**
## Summary

`jax.scipy.special.digamma(0.0)` currently returns `NaN` because the underlying `lax.digamma` kernel returns `NaN` at the simple pole at the origin. `scipy.special.digamma(0.0)` returns `-inf` (and `+inf` for `-0.0`, the signed limit from each side of the pole). This PR wraps the `lax.digamma` call with a `jnp.where` guard so the JAX path matches scipy parity.

The mask-to-safe-value pattern (used elsewhere in this file, e.g. `_loggamma_lanczos`) keeps gradients through the unselected path free of NaN. All other values are unchanged.

### Before
```
digamma(0.0):  scipy=-inf   jax=nan    wrong
digamma(-0.0): scipy=+inf   jax=nan    wrong
digamma(0.5):  scipy=-1.96  jax=-1.96  ok
```

### After
```
digamma(0.0):  scipy=-inf   jax=-inf   ok
digamma(-0.0): scipy=+inf   jax=+inf   ok
digamma(0.5):  scipy=-1.96  jax=-1.96  ok
```

Note: the secondary subnormal-input issue called out in the bug report (subnormals being flushed to zero by XLA's DAZ mode) is a consequence of `digamma(0)` returning `NaN`, so this fix resolves that path too on backends where the DAZ flush routes subnormals through the new branch.

Fixes #37763

## Test plan

- [x] New regression test `testDigammaBoundaryValues` added to `tests/lax_scipy_special_functions_test.py` that exercises `0.0`, `-0.0`, `1.0`, `0.5` via `_CheckAgainstNumpy` and `_CompileAndCheck`.
- [x] Existing digamma tests (`testScipySpecialFun_digamma_*`, including the autodiff tests) still pass.
- [x] Full `lax_scipy_special_functions_test.py` suite (370 tests) passes locally.
- [x] `ruff check` clean on the touched files.

### CODE REVIEW

**Comment by adityasingh2400** [comment]
@googlebot I signed it!

**Comment by adityasingh2400** [comment]
Applied both gemini-code-assist suggestions in 83af44d.

Scalar fusion in special.py. Switched x_safe and pole_val to scalar fills with .astype(x.dtype) preserved so the float32 and float64 paths stay correct under jax_enable_x64.

Added -1.0 to x_samples. Verified scipy.special.digamma(-1.0) returns NaN for both float32 and float64, and jax.lax.digamma(-1.0) returns NaN as well, so _CheckAgainstNumpy matches via assertAllClose NaN-equality. Ran a local parity check across [0.0, -0.0, 1.0, 0.5, -1.0] for both dtypes and JAX matches scipy on every sample.

**Comment by jakevdp** [comment]
This change may be fine, but if we want this behavior I think it would be better to fix the StableHLO `digamma` implementation itself. It looks like the code explicitly returns `nan` rather than `inf` at the poles of the function: https://github.com/openxla/stablehlo/blob/effc2e60751b8e900f11c612c3f5b07e2a5bbfd4/stablehlo/transforms/ChloLegalizeToStablehlo.cpp#L1635

On the other hand, the existing semantics are arguably more correct, because the value diverges to negative or positive infinity depending on whether you approach the limit from the left or from the right. Given this, the response from StableHLO maintainers may be "works as intended".

**Comment by adityasingh2400** [comment]
Good points, and I think you have convinced me the venue is wrong.

You are right that the pole is genuinely two-sided: digamma diverges to negative infinity approaching 0 from the right and positive infinity from the left, so NaN is a defensible "the limit does not exist" answer. The only argument for the change is parity, since scipy picks the right-hand limit and returns -inf at x=0, but that is scipy making an arbitrary choice rather than a mathematically forced one.

Given that, patching the result in the JAX wrapper is the wrong layer. If we want the scipy convention it belongs in the StableHLO digamma lowering you linked, and as you note the StableHLO maintainers may reasonably call the current NaN working as intended.

I am happy to close this and instead open an issue on the StableHLO side to ask whether they want to match scipy at the poles, unless you would rather keep this open as a tracking reference. Let me know which you prefer.

**Comment by adityasingh2400** [comment]
Closing this out. Since I last commented I am more convinced the NaN is defensible on its own terms: the pole is two-sided, so "the limit does not exist" is a reasonable answer and scipy returning -inf is scipy picking the right-hand limit by convention rather than by necessity. Patching it in the JAX wrapper would paper over the StableHLO lowering, and pushing it upstream would likely get a working-as-intended response for the same reason.

Not worth more of the team's review time either way, so I am closing rather than leaving it open in the queue. Thanks for the careful read on this one.
